plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java

```diff
@@ -0,0 +1,419 @@
+        statement.setObjectName("statement");
+        return statement;
+    }
+
+    @Override
+    public List<QuotaTariffVO> listQuotaTariffPlans(final QuotaTariffListCmd cmd) {
+        List<QuotaTariffVO> result = new ArrayList<QuotaTariffVO>();
+        Date effectiveDate = cmd.getEffectiveDate() == null ? new Date() : cmd.getEffectiveDate();
+        Date adjustedEffectiveDate = _quotaService.computeAdjustedTime(effectiveDate);
+        s_logger.info("Effective datec=" + effectiveDate + " quotatype=" + cmd.getUsageType() + " Adjusted date=" + adjustedEffectiveDate);
+        if (cmd.getUsageType() != null) {
+            QuotaTariffVO tariffPlan = _quotaTariffDao.findTariffPlanByUsageType(cmd.getUsageType(), adjustedEffectiveDate);
+            if (tariffPlan != null) {
+                result.add(tariffPlan);
+            }
+        } else {
+            result = _quotaTariffDao.listAllTariffPlans(adjustedEffectiveDate);
+        }
+        return result;
+    }
+
+    @Override
+    public QuotaTariffVO updateQuotaTariffPlan(QuotaTariffUpdateCmd cmd) {
+        final int quotaType = cmd.getUsageType();
+        final BigDecimal quotaCost = new BigDecimal(cmd.getValue());
+        final Date effectiveDate = _quotaService.computeAdjustedTime(cmd.getStartDate());
+        final Date now = _quotaService.computeAdjustedTime(new Date());
+        // if effective date is in the past return error
+        if (effectiveDate.compareTo(now) < 0) {
+            throw new InvalidParameterValueException("Incorrect effective date for tariff " + effectiveDate + " is less than now " + now);
+        }
+        QuotaTypes quotaConstant = QuotaTypes.listQuotaTypes().get(quotaType);
+        if (quotaConstant == null) {
+            throw new InvalidParameterValueException("Quota type does not exists " + quotaType);
+        }
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.USAGE_DB).close();
+        QuotaTariffVO result = null;
+        try {
+            result = new QuotaTariffVO();
+            result.setUsageType(quotaType);
+            result.setUsageName(quotaConstant.getQuotaName());
+            result.setUsageUnit(quotaConstant.getQuotaUnit());
+            result.setUsageDiscriminator(quotaConstant.getDiscriminator());
+            result.setCurrencyValue(quotaCost);
+            result.setEffectiveOn(effectiveDate);
+            result.setUpdatedOn(now);
+            result.setUpdatedBy(cmd.getEntityOwnerId());
+
+            s_logger.debug(String.format("Updating Quota Tariff Plan: New value=%s for resource type=%d effective on date=%s", quotaCost, quotaType, effectiveDate));
+            _quotaTariffDao.addQuotaTariff(result);
+        } catch (Exception pokemon) {
+            s_logger.error("Error in update quota tariff plan: " + pokemon);
+        } finally {
+            TransactionLegacy.open(opendb).close();
```

Comment:
Per comments above, the immediate open and close of a database connection needs to be replaced with the newer transaction management mechanism.